runtime/src/commonMain/kotlin/pbandk/FieldDescriptor.kt

```diff
@@ -125,4 +132,52 @@ class FieldDescriptor<M : Message, T>(
             override fun isDefaultValue(value: Any?) = (value as? kotlin.collections.Map<*, *>)?.isEmpty() == true
         }
     }
+
+    @Suppress("UNCHECKED_CAST")
+    fun <M : Message, T> parseUnknownField(
+        unknownField: UnknownField.Value,
+        fieldDescriptor: FieldDescriptor<M, T>
+    ): T {
+        return when (fieldDescriptor.type) {
+            is Type.Primitive<*> -> parseUnknownFieldForPrimitiveOrComposite(fieldDescriptor.type, unknownField) as T
+            is Type.Message<*> -> parseUnknownFieldForTypeMessage(fieldDescriptor.type, unknownField) as T
+            else -> throw UnsupportedOperationException("fieldDescriptor.type: '${fieldDescriptor.type}' is not supported")
+        }
+    }
+
+    private fun parseUnknownFieldForPrimitiveOrComposite(type: Type.Primitive<*>, unknownField: UnknownField.Value): T {
+        return if (unknownField is UnknownField.Value.Composite) {
+            parseUnknownFieldForPrimitive(type, unknownField.values.last())
+        } else {
+            parseUnknownFieldForPrimitive(type, unknownField)
+        }
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    private fun parseUnknownFieldForPrimitive(type: Type.Primitive<*>, unknownField: UnknownField.Value) : T {
+       return when (type) {
```

Comment:
There are a lot of types here and I wasn't sure if I should be supporting all of them so their currently missing.

Reply:
Yeah we need to support all of them, since a custom option can use any of the types.